### PR TITLE
feat(sqlsmith): gen inserts

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Implements `GEN INSERTs` (#3844)
- Disabled for now, as there's still outstanding bug: #8059 
- Implement a basic query extractor for deterministic test logs.
- Added golden test for this (hence the large diff).
- Reduced `CROSS JOINs` to one, else materialized view takes a long time to run.
- Reduced occurrence of expressions resulting in stream nested loop join.
- Increased chance to select columns rather than scalars.
- Further improvements can be done in other PR.

## Checklist

- [x] Throttle inserts to 10 / table first.
- [x] Additionally, it should insert into every table.
- [x] Log inserts for reproducibility.
~- [ ] Generate more `JOIN ON`  tables.~
~- [ ] Add some non_equi joins.~
- [x] Use simple_query protocol to avoid binary format. #8039 .
- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/tests/sqlsmith/src/runner.rs [hunk 1]
```diff
@@ -13,21 +13,44 @@
...
+    tracing::info!("Created tables");
+
+    let row_count = 500;
+    client
+        .query("SET RW_IMPLICIT_FLUSH TO TRUE;", &[])
+        .await
+        .unwrap();
+    populate_tables(client, &mut rng, base_tables.clone(), row_count).await;
+    tracing::info!("Created base_tables");
+    // FIXME(noel): This is a hack, seems like flush is not working as expected.
```
Comment (kwannoel) [inline src/tests/sqlsmith/src/runner.rs:42]: Fix this, it's duplicate values on pk

File: src/tests/sqlsmith/src/sql_gen/insert.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        Statement::Insert {
+            table_name,
+            columns: vec![],
+            source: Box::new(source),
+            returning: vec![],
+        }
+    }
+
+    fn gen_values(&mut self, data_types: &[DataType], row_count: usize) -> Vec<Vec<Expr>> {
+        (0..row_count).map(|_| self.gen_row(data_types)).collect()
```
Comment (jon-chuang) [inline src/tests/sqlsmith/src/sql_gen/insert.rs:51]: Will `gen_row` also generate `NULL`s?

Comment (jon-chuang) [reply src/tests/sqlsmith/src/sql_gen/insert.rs:51]: Maybe let's add a small chance of generating NULL

File: src/tests/sqlsmith/src/runner.rs [hunk 3]
```diff
@@ -136,16 +193,18 @@ async fn create_tables(
...
     let mut setup_sql = String::with_capacity(1000);
     let sql = get_seed_table_sql(testdata);
     let statements = parse_sql(&sql);
-    let mut tables = statements
+    let mut tables = vec![];
+    let base_tables = statements
         .iter()
         .map(create_table_statement_to_table)
         .collect_vec();
+    tables.extend_from_slice(&base_tables);
```
Comment (jon-chuang) [inline src/tests/sqlsmith/src/runner.rs:207]: mabe just write `mut tables = base_tables.clone()`

Comment (jon-chuang) [reply src/tests/sqlsmith/src/runner.rs:207]: Maybe rename it to `mvs_and_base_tables` for clarity...

File: src/tests/sqlsmith/src/sql_gen/insert.rs (hunk 2 above)
Comment (kwannoel) [reply src/tests/sqlsmith/src/sql_gen/insert.rs:51]: `gen_simple_scalar` can generate nulls. But keep it disabled, there's still a related open issue. #7327

File: src/tests/sqlsmith/src/runner.rs [hunk 4]
```diff
@@ -37,13 +55,31 @@ pub async fn run(client: &tokio_postgres::Client, testdata: &str, count: usize)
...
 /// Sanity checks for sqlsmith
 async fn test_sqlsmith<R: Rng>(
     client: &tokio_postgres::Client,
     rng: &mut R,
     tables: Vec<Table>,
     setup_sql: &str,
+    base_tables: Vec<Table>,
+    row_count: usize,
 ) {
+    // Test inserted rows should be at least 50% population count,
```
Comment (fuyufjh) [inline src/tests/sqlsmith/src/runner.rs:78]: Is this because that <=50% inserted rows are duplicated in PK?

Comment (kwannoel) [reply src/tests/sqlsmith/src/runner.rs:78]: Yes, otherwise it will affect effectiveness of sqlsmith

**Review by fuyufjh — APPROVED** [review]

**Comment by kwannoel** [comment]
Need to gen struct and array in binary format... working on it.

**Comment by kwannoel** [comment]
New bugs found by sqlsmith 🥲

**Comment by kwannoel** [comment]
Mark as ready-for-review to run on ci first. Will run it for a few days to see if it is stable.

**Comment by kwannoel** [comment]
- Disable gen inserts for now to get it merged first, since there's a bug outstanding, and don't want this PR to get larger.
- Further changes still needed:
   - [ ] More equi joins (in-place of removed CROSS JOINs).
   - [ ] Basic shrinker.
- Will work on these in a separate PR.

File: scripts/sqlsmith/extract_failing_query.sh [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+# assumes it is last
+extract_query() {
+    tail -n 1 | extract_sql
+}
+
+########## EXTRACT QUERY + DDL + DML
+
+if [ "$1" = "--pretty" ]; then
```
Comment (kwannoel) [inline scripts/sqlsmith/extract_failing_query.sh:38]: Will refactor and use `getopts` in a later PR. Planning to add basic shrinker here.

File: src/common/src/array/column_proto_readers.rs [hunk 6]
```diff
@@ -68,7 +68,12 @@ pub fn read_bool_array(array: &ProstArray, cardinality: usize) -> ArrayResult<Ar
     let bitmap: Bitmap = array.get_null_bitmap()?.into();
 
     let arr = BoolArray::new(data, bitmap);
-    assert_eq!(arr.len(), cardinality);
+    // Either array length should match cardinality,
+    // or it is null, and array length should be 0.
+    // To avoid confusion, "len" here does not refer to length of array,
+    // rather, it is the number of rows of arrays returned.
+    // cardinality also refers to this.
+    assert!(arr.len() == cardinality || arr.len() == 0);
```
Comment (kwannoel) [inline src/common/src/array/column_proto_readers.rs:76]: cc @xiangjinwu is there a better assertion we can make for `null` case?

`arr.len() == 0` doesn't seem very precise...

Comment (kwannoel) [reply src/common/src/array/column_proto_readers.rs:76]: Added a e2e test case for this. Previously it panics on this assertion, see e2e test case for concrete sql.

**Comment by kwannoel** [comment]
Failed due to `batch`, seems unrelated. Re-running.

File: src/common/src/array/column_proto_readers.rs (hunk 6 above)
Comment (xiangjinwu) [reply src/common/src/array/column_proto_readers.rs:76]: This looks like a bug in the caller or even sender encoding side. I am taking a look.

Comment (xiangjinwu) [reply src/common/src/array/column_proto_readers.rs]: #8082 done

**Review by lmatz — APPROVED** [review]
LGTM

**Pull request merged**